File: Backtracking/GenerateParentheses.java
// LeetCode 22 - Generate Parentheses
// Time Complexity: O(4^n / sqrt(n)) | Space Complexity: O(n)
import java.util.ArrayList;
import java.util.List;

public class GenerateParentheses {
    public List<String> generateParenthesis(int n) {
        List<String> answer = new ArrayList<>();
        dfs(n, 0, 0, new StringBuilder(), answer);
        return answer;
    }

    private void dfs(int n, int open, int close, StringBuilder path, List<String> answer) {
        if (open == n && close == n) {
            answer.add(path.toString());
            return;
        }

        if (open < n) {
            path.append('(');
            dfs(n, open + 1, close, path, answer);
            path.deleteCharAt(path.length() - 1);
        }

        if (close < open) {
            path.append(')');
            dfs(n, open, close + 1, path, answer);
            path.deleteCharAt(path.length() - 1);
        }
    }
}

File: Backtracking/GenerateParentheses.go
// LeetCode 22 - Generate Parentheses
// Time Complexity: O(4^n / sqrt(n)) | Space Complexity: O(n)
package backtracking

func generateParenthesis(n int) []string {
	ans := []string{}

	var dfs func(int, int, []byte)
	dfs = func(open int, close int, path []byte) {
		if open == n && close == n {
			ans = append(ans, string(path))
			return
		}

		if open < n {
			dfs(open+1, close, append(path, '('))
		}

		if close < open {
			dfs(open, close+1, append(path, ')'))
		}
	}

	dfs(0, 0, []byte{})
	return ans
}
